File: src/main/java/com/anod/appwatcher/ui/SettingsActionBarActivity.java
package com.anod.appwatcher.ui;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.LayoutRes;
import android.support.annotation.StringRes;
import android.support.v7.widget.Toolbar;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.CheckBox;
import android.widget.ListView;
import android.widget.TextView;

import com.anod.appwatcher.R;
import com.anod.appwatcher.utils.MenuItemAnimation;

import java.util.ArrayList;
import java.util.List;

abstract public class SettingsActionBarActivity extends ToolbarActivity implements AdapterView.OnItemClickListener {
    protected ListView mListView;
    private PreferenceAdapter mPreferenceAdapter;
    private MenuItemAnimation mRefreshAnim;

    public static class Preference {
        final int title;
        final int layout;
        public Preference(@StringRes int title,@LayoutRes int layout) {
            this.title = title;
            this.layout = layout;
        }
    }
    public static class Category extends Preference {
        public Category(@StringRes int title) {
            super(title, R.layout.preference_category);
        }
    }

    public static class Item extends Preference {
        final int action;
        public int summaryRes;
        public String summary;
        public int widget;
        public boolean enabled = true;

        public Item(@StringRes int title,@StringRes int summaryRes, int action) {
            super(title, R.layout.preference_holo);
            this.summaryRes = summaryRes;
            this.action = action;
        }
    }

    public static class CheckboxItem extends Item {
        public boolean checked = false;

        public CheckboxItem(@StringRes int title,@StringRes int summaryRes, int action) {
            super(title, summaryRes, action);
            this.widget = R.layout.preference_widget_checkbox;
        }

        public CheckboxItem(int title, int summaryRes, int action, boolean checked) {
            this(title, summaryRes, action);
            this.checked = checked;
        }

        public void switchState() {
            this.checked = !this.checked;
        }
    }

    static class PreferenceAdapter extends ArrayAdapter<Preference> {

        private final SettingsActionBarActivity mActivity;
        private final LayoutInflater mInflater;

        public PreferenceAdapter(SettingsActionBarActivity activity, List<Preference> objects) {
            super(activity, 0, objects);
            mActivity = activity;
            mInflater = (LayoutInflater)mActivity.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        }

        @Override
        public int getItemViewType(int position) {
            Preference pref = getItem(position);
            if (pref instanceof CheckboxItem) {
                return 0;
            }
            if (pref instanceof Category) {
                return 1;
            }
            return 2;
        }

        @Override
        public int getViewTypeCount() {
            return 3;
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            final Preference pref = getItem(position);

            View view;
            if (convertView == null) {
                view = mInflater.inflate(pref.layout, parent, false);
            } else {
                view = convertView;
            }

            TextView title = (TextView) view.findViewById(android.R.id.title);
            title.setText(pref.title);

            if (pref instanceof Item) {
                Item item = (Item) pref;
                View icon = view.findViewById(android.R.id.icon);
                if (icon != null) {
                    icon.setVisibility(View.GONE);
                }

                TextView summary = (TextView) view.findViewById(android.R.id.summary);
                if (item.summaryRes > 0) {
                    summary.setText(item.summaryRes);
                } else if (!TextUtils.isEmpty(item.summary)) {
                    summary.setText(item.summary);
                }

                final ViewGroup widgetFrame = (ViewGroup) view.findViewById(android.R.id.widget_frame);
                if (item.widget > 0) {

                    if (item instanceof CheckboxItem) {
                        CheckBox checkBox = (CheckBox) widgetFrame.findViewById(android.R.id.checkbox);
                        if (checkBox == null) {
                            mInflater.inflate(item.widget, widgetFrame);
                            checkBox = (CheckBox) widgetFrame.findViewById(android.R.id.checkbox);
                        }
                        checkBox.setChecked(((CheckboxItem) item).checked);
                    }
                } else {
                    widgetFrame.setVisibility(View.GONE);
                }
            }

            return view;
        }

        @Override
        public boolean isEnabled(int position) {
            final Preference pref = getItem(position);
            if(pref instanceof Category) {
                return false;
            }
            if (pref instanceof Item) {
                return ((Item) pref).enabled;
            }
            return true;
        }
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_settings);
        setupToolbar();

        mRefreshAnim = new MenuItemAnimation(this, R.anim.rotate);
        mRefreshAnim.setInvisibleMode(true);
        init();

        ArrayList<Preference> preferences = initPreferenceItems();

        mPreferenceAdapter = new PreferenceAdapter(this, preferences);
        mListView = (ListView) findViewById(android.R.id.list);
        mListView.setEmptyView(findViewById(android.R.id.empty));
        mListView.setAdapter(mPreferenceAdapter);
        mListView.setOnItemClickListener(this);
    }

    protected abstract void init();
    protected abstract ArrayList<Preference> initPreferenceItems();
    protected abstract void onPreferenceItemClick(int action, Item pref);

    protected void setProgressVisibility(boolean visible) {
        if (visible) {
            mRefreshAnim.start();
        } else {
            mRefreshAnim.stop();
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu
        getMenuInflater().inflate(R.menu.settings, menu);
        MenuItem refreshMenuItem = menu.findItem(R.id.menu_act_refresh);
        mRefreshAnim.setMenuItem(refreshMenuItem);

        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
        Preference pref = (Preference) mListView.getItemAtPosition(position);
        if (pref instanceof Item) {
            int action = ((Item) pref).action;
            if (pref instanceof CheckboxItem) {
                ((CheckboxItem) pref).switchState();
            }
            onPreferenceItemClick(action, (Item) pref);
        }
    }

    protected void notifyDataSetChanged() {
        mPreferenceAdapter.notifyDataSetChanged();
    }

}

File: src/main/java/com/anod/appwatcher/ui/SettingsActionBarActivity.kt
package com.anod.appwatcher.ui

import android.content.Context
import android.os.Bundle
import android.support.annotation.LayoutRes
import android.support.annotation.StringRes
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.Menu
import android.view.View
import android.view.ViewGroup
import android.widget.*
import com.anod.appwatcher.R
import com.anod.appwatcher.utils.MenuItemAnimation

abstract class SettingsActionBarActivity : ToolbarActivity(), AdapterView.OnItemClickListener {
    lateinit private var mListView: ListView
    lateinit private var mRefreshAnim: MenuItemAnimation
    lateinit private var mPreferenceAdapter: PreferenceAdapter

    open class Preference(@StringRes val title: Int, @LayoutRes val layout: Int)
    class Category(@StringRes title: Int) : Preference(title, R.layout.preference_category)

    open class Item(@StringRes title: Int, @StringRes var summaryRes: Int, internal val action: Int)
        : Preference(title, R.layout.preference_holo) {
        var summary: String? = null
        var widget: Int = 0
        var enabled = true
    }

    class CheckboxItem(@StringRes title: Int, @StringRes summaryRes: Int, action: Int)
        : Item(title, summaryRes, action) {
        var checked = false

        init {
            this.widget = R.layout.preference_widget_checkbox
        }

        constructor(title: Int, summaryRes: Int, action: Int, checked: Boolean) : this(title, summaryRes, action) {
            this.checked = checked
        }

        fun switchState() {
            this.checked = !this.checked
        }
    }

    internal class PreferenceAdapter(activity: SettingsActionBarActivity, objects: List<Preference>) : ArrayAdapter<Preference>(activity, 0, objects) {
        private val mInflater: LayoutInflater = activity.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater

        override fun getItemViewType(position: Int): Int {
            val pref = getItem(position)
            if (pref is CheckboxItem) {
                return 0
            }
            if (pref is Category) {
                return 1
            }
            return 2
        }

        override fun getViewTypeCount(): Int {
            return 3
        }

        override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
            val pref = getItem(position)

            val view: View
            if (convertView == null) {
                view = mInflater.inflate(pref!!.layout, parent, false)
            } else {
                view = convertView
            }

            val title = view.findViewById<View>(android.R.id.title) as TextView
            title.setText(pref!!.title)

            if (pref is Item) {
                val item = pref
                val icon = view.findViewById<View>(android.R.id.icon)
                if (icon != null) {
                    icon.visibility = View.GONE
                }

                val summary = view.findViewById<View>(android.R.id.summary) as TextView
                if (item.summaryRes > 0) {
                    summary.setText(item.summaryRes)
                } else if (!TextUtils.isEmpty(item.summary)) {
                    summary.text = item.summary
                }

                val widgetFrame = view.findViewById<View>(android.R.id.widget_frame) as ViewGroup
                if (item.widget > 0) {

                    if (item is CheckboxItem) {
                        var checkBox: CheckBox? = widgetFrame.findViewById<View>(android.R.id.checkbox) as? CheckBox
                        if (checkBox == null) {
                            mInflater.inflate(item.widget, widgetFrame)
                            checkBox = widgetFrame.findViewById<View>(android.R.id.checkbox) as CheckBox
                        }
                        checkBox.isChecked = item.checked
                    }
                } else {
                    widgetFrame.visibility = View.GONE
                }
            }

            return view
        }

        override fun isEnabled(position: Int): Boolean {
            val pref = getItem(position)
            if (pref is Category) {
                return false
            }
            if (pref is Item) {
                return pref.enabled
            }
            return true
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_settings)
        setupToolbar()

        mRefreshAnim = MenuItemAnimation(this, R.anim.rotate)
        mRefreshAnim.isInvisibleMode = true
        init()

        val preferences = initPreferenceItems()

        mPreferenceAdapter = PreferenceAdapter(this, preferences)
        mListView = findViewById<View>(android.R.id.list) as ListView
        mListView.emptyView = findViewById(android.R.id.empty)
        mListView.adapter = mPreferenceAdapter
        mListView.onItemClickListener = this
    }

    protected abstract fun init()
    protected abstract fun initPreferenceItems(): List<Preference>
    protected abstract fun onPreferenceItemClick(action: Int, pref: Item)

    protected fun setProgressVisibility(visible: Boolean) {
        if (visible) {
            mRefreshAnim.start()
        } else {
            mRefreshAnim.stop()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu
        menuInflater.inflate(R.menu.settings, menu)
        mRefreshAnim.menuItem = menu.findItem(R.id.menu_act_refresh)

        return super.onCreateOptionsMenu(menu)
    }

    override fun onItemClick(parent: AdapterView<*>, view: View, position: Int, id: Long) {
        val pref = mListView.getItemAtPosition(position) as Preference
        if (pref is Item) {
            val action = pref.action
            if (pref is CheckboxItem) {
                pref.switchState()
            }
            onPreferenceItemClick(action, pref)
        }
    }

    protected fun notifyDataSetChanged() {
        mPreferenceAdapter.notifyDataSetChanged()
    }

}
